### PR TITLE
#patch: Add in task execution event fields

### PR DESCRIPTION
**Pull request by katrogan**
Signed-off-by: Katrina Rogan <katroganGH@gmail.com>

# TL;DR
Add in task execution event fields

- Add **phase reason** and **task type** to TaskExecutionEvent
- Add **generated_name** to TaskExecutionMetadata
- Add a **ManagedResourceInfo** message for capturing resource token name (and future FRM attributes)

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/325

## Follow-up issue
_NA_

### CODE REVIEW

File: protos/flyteidl/event/event.proto [hunk 1]
```diff
@@ -167,4 +185,10 @@ message TaskExecutionMetadata {
         INTERRUPTIBLE = 1;
     }
     InstanceClass instance_class = 16;
+
+    // Generated unique name for this task execution used by the backend.
+    string generated_name = 1;
+
+    // Includes information about how resource token allocation (if applicable).
+    ManagedResourceInfo managed_resource_info = 2;
```
Comment (EngHabu) [inline protos/flyteidl/event/event.proto:193]: should this be repeated?

File: protos/flyteidl/event/event.proto [hunk 2]
```diff
@@ -151,12 +151,30 @@ message TaskExecutionEvent {
...
+// This message holds task execution metadata specific to resource allocation used to manage concurrent
+// executions for a project namespace.
+message ManagedResourceInfo {
+    // Unique resource ID used to identify this execution when allocating a token.
+    string token                         = 1;
+}
+
 // Holds metadata around how a task was executed.
 // TODO(katrogan): Extend to include freeform fields (https://github.com/flyteorg/flyte/issues/325).
+// Next ID to use:  3
```
Comment (EngHabu) [inline protos/flyteidl/event/event.proto:177]: While I liked these back in the day, unless we have tools to make sure they are correct (do we?) they will get out of sync and be confusing... I think.

File: protos/flyteidl/event/event.proto [hunk 3]
```diff
@@ -151,12 +151,30 @@ message TaskExecutionEvent {
...
     uint32 phase_version                  = 12;    
 
+    // If there is an explanation for this phase transition, the reason will capture it.
+    string reason                         = 13;
+
+
+    // A predefined yet extensible Task type identifier. If the task definition is already registered in flyte admin
+    // this type will be identical, but not all task executions necessarily use pre-registered definitions and this
+    // type is useful to render the task in the UI, filter task executions, etc.
+    string type                           = 14;
```
Comment (EngHabu) [inline protos/flyteidl/event/event.proto:161]: Is this the task type? or the event type? from the comment it seems it refers to the task type... maybe it should be renamed `task_type` ?

Comment (katrogan) [reply protos/flyteidl/event/event.proto:161]: ah good point, done

File: protos/flyteidl/event/event.proto (hunk 2 above)
Comment (katrogan) [reply protos/flyteidl/event/event.proto]: fair point, done

File: protos/flyteidl/event/event.proto (hunk 1 above)
Comment (katrogan) [reply protos/flyteidl/event/event.proto:192]: perhaps i'm misunderstanding but isn't this only determined once per execution?

Comment (EngHabu) [reply protos/flyteidl/event/event.proto:192]: A plugin can request multiple resource allocations during execution... our plugins only request one resource at most because we didn't need more... but the API allows plugins to request more

Comment (katrogan) [reply protos/flyteidl/event/event.proto]: @EngHabu got it, thanks. updated

File: protos/flyteidl/event/event.proto [hunk 4]
```diff
@@ -151,10 +151,27 @@ message TaskExecutionEvent {
...
     // Metadata around how a task was executed.
     TaskExecutionMetadata metadata        = 16;
 }
 
+
+// This message holds task execution metadata specific to resource allocation used to manage concurrent
+// executions for a project namespace.
+message ManagedResourceInfo {
+    // Unique resource ID used to identify this execution when allocating a token.
+    string token                         = 1;
```
Comment (EngHabu) [inline protos/flyteidl/event/event.proto:172]: Should we call this `allocation_token` ?

Comment (EngHabu) [reply protos/flyteidl/event/event.proto:172]: Should we capture more of the request? https://github.com/flyteorg/flyteplugins/blob/master/go/tasks/pluginmachinery/core/resource_manager.go#L82
the resource namespace maybe?

**Review by EngHabu — DISMISSED** [review]

File: protos/flyteidl/event/event.proto (hunk 4 above)
Comment (katrogan) [reply protos/flyteidl/event/event.proto]: sure done @EngHabu

**Review by EngHabu — DISMISSED** [review]

File: protos/flyteidl/event/event.proto [hunk 5]
```diff
@@ -167,4 +187,11 @@ message TaskExecutionMetadata {
         INTERRUPTIBLE = 1;
     }
     InstanceClass instance_class = 16;
+
+    // Generated unique name for this task execution used by the backend.
+    string generated_name = 1;
```
Comment (kumare3) [inline protos/flyteidl/event/event.proto:192]: should we call it resource_name?
As for Qubole/AWS Batch etc, it may not be the generated_name?
Or we can have both - but that is confusing!

Comment (katrogan) [reply protos/flyteidl/event/event.proto]: done

File: protos/flyteidl/event/event.proto [hunk 6]
```diff
@@ -151,10 +151,40 @@ message TaskExecutionEvent {
     // The version field should be incremented when metadata changes across the duration of an individual phase.
     uint32 phase_version                  = 12;    
 
+    // If there is an explanation for this phase transition, the reason will capture it.
```
Comment (kumare3) [inline protos/flyteidl/event/event.proto:154]:
```suggestion
    // An optional explanation for the phase transition
```

File: protos/flyteidl/event/event.proto [hunk 7]
```diff
@@ -151,10 +151,40 @@ message TaskExecutionEvent {
...
+    string reason                         = 13;
+
+
+    // A predefined yet extensible Task type identifier. If the task definition is already registered in flyte admin
+    // this type will be identical, but not all task executions necessarily use pre-registered definitions and this
+    // type is useful to render the task in the UI, filter task executions, etc.
+    string task_type                      = 14;
```
Comment (kumare3) [inline protos/flyteidl/event/event.proto:161]: should we also add the actual execution plugin-ID used?

File: protos/flyteidl/event/event.proto [hunk 8]
```diff
@@ -167,4 +197,14 @@ message TaskExecutionMetadata {
         INTERRUPTIBLE = 1;
     }
     InstanceClass instance_class = 16;
+
+    // Contains various identifiers for resources used during execution.
```
Comment (kumare3) [inline protos/flyteidl/event/event.proto:201]: nit: can we order them by the field id? sorry!

Comment (katrogan) [reply protos/flyteidl/event/event.proto:192]: @kumare3 done

File: protos/flyteidl/event/event.proto (hunk 6 above)
Comment (katrogan) [reply protos/flyteidl/event/event.proto]: done

**Comment by katrogan** [comment]
friendly ping @EngHabu @kumare3

File: protos/flyteidl/event/event.proto [hunk 9]
```diff
@@ -151,13 +151,54 @@ message TaskExecutionEvent {
...
+    string allocation_token               = 1;
+
+    // Namespace under which this task execution requested an allocation token.
+    string namespace                      = 2;
+}
+
 // Holds metadata around how a task was executed.
 // TODO(katrogan): Extend to include freeform fields (https://github.com/flyteorg/flyte/issues/325).
 message TaskExecutionMetadata {
+
```
Comment (kumare3) [inline protos/flyteidl/event/event.proto:188]: Can we remove the todo on the message? indicates task -
https://github.com/flyteorg/flyte/issues/325

File: protos/flyteidl/event/event.proto [hunk 10]
```diff
@@ -151,13 +151,54 @@ message TaskExecutionEvent {
...
 
+    // An optional explanation for the phase transition.
+    string reason                         = 13;
+
+
+    // A predefined yet extensible Task type identifier. If the task definition is already registered in flyte admin
+    // this type will be identical, but not all task executions necessarily use pre-registered definitions and this
+    // type is useful to render the task in the UI, filter task executions, etc.
+    string task_type                      = 14;
+
```
Comment (kumare3) [inline protos/flyteidl/event/event.proto:162]: @katrogan another question, why is task_type / reason not metadata? Maybe a clearer definition in the distinction

File: protos/flyteidl/event/event.proto [hunk 11]
```diff
@@ -151,13 +151,54 @@ message TaskExecutionEvent {
...
     // Metadata around how a task was executed.
     TaskExecutionMetadata metadata        = 16;
 }
 
+// This message contains metadata about external resources produced or used by a specific task execution.
+message ExternalResourceInfo {
+
+    // Identifiers for external resources created by this task execution, for example Qubole query ID or presto query ids.
+    repeated string external_ids = 1;
```
Comment (EngHabu) [inline protos/flyteidl/event/event.proto:171]: If we have a repeated field [here](https://github.com/flyteorg/flyteidl/pull/119/files#diff-61c99ca77e3f94bf6d64036e11e8851dd5a9defc8fb5f1d385af4bfc52b9e9f4R197) we don't need this to be repeated, right? this can include other things if the plugin wants to (maybe cost related, maybe remote cluster/region info ... etc.) that we can open up later

File: protos/flyteidl/event/event.proto (hunk 10 above)
Comment (katrogan) [reply protos/flyteidl/event/event.proto:162]: So I like the documentation at the beginning of this proto message: `Plugin specific execution event information`  i.e. attributes of the event itself. In this case task_type is an attribute of the execution and reason goes along with phase and phase version as part of the actual event update, right? I see metadata is more for identifying/discerning how a task was executed rather than artifacts or products of the event itself.

File: protos/flyteidl/event/event.proto (hunk 9 above)
Comment (katrogan) [reply protos/flyteidl/event/event.proto:187]: oh yes good catch, done

File: protos/flyteidl/event/event.proto (hunk 11 above)
Comment (katrogan) [reply protos/flyteidl/event/event.proto]: done

**Review by kumare3 — DISMISSED** [review]

**Comment by katrogan** [comment]
one last (?) ping @kumare3 @EngHabu

**Review by kumare3 — APPROVED** [review]

**Review by EngHabu — APPROVED** [review]
👍
